// Copyright 2022 Silent Eight Pte. Ltd.

syntax = "proto3";

package silenteight.datascience.etlpipeline.v1.api;

import "google/api/annotations.proto";
import "google/api/field_behavior.proto";

option java_package = "com.silenteight.datascience.etlpipeline.v1.api";
option java_outer_classname = "EtlPipelineApiProto";
option java_multiple_files = true;

service EtlPipelineService {

  rpc RunEtl (RunEtlRequest) returns (RunEtlResponse) {
    option (google.api.http) = {
      post: "/v1/etl:runEtl"
      body: "runEtl"
    };
  }
}

message RunEtlRequest {
  repeated Alert alerts = 1 [(google.api.field_behavior) = REQUIRED];
}

message Alert {
  string batch_id = 1 [(google.api.field_behavior) = REQUIRED];
  string alert_name = 2 [(google.api.field_behavior) = REQUIRED];
  repeated Match matches = 3 [(google.api.field_behavior) = REQUIRED];
  map<string, string> flat_payload = 4 [(google.api.field_behavior) = REQUIRED];
}

message Match {
  string match_id = 1 [(google.api.field_behavior) = REQUIRED];
  string match_name = 2 [(google.api.field_behavior) = REQUIRED];
}

message RunEtlResponse {
  repeated EtlAlert etl_alerts = 1 [(google.api.field_behavior) = REQUIRED];
}

message EtlAlert {
  string batch_id = 1 [(google.api.field_behavior) = REQUIRED];
  string alert_name = 2 [(google.api.field_behavior) = REQUIRED];
  EtlStatus etl_status = 3 [(google.api.field_behavior) = REQUIRED];
  repeated EtlMatch etl_matches = 4 [(google.api.field_behavior) = REQUIRED];
}

message EtlMatch {
  string match_name = 1 [(google.api.field_behavior) = REQUIRED];
}

enum EtlStatus {
  UNKNOWN = 0;
  FAILURE = 1;
  SUCCESS = 2;
}
